experimental/QuadFormAndIsom/src/spaces_with_isometry.jl

```diff
@@ -0,0 +1,796 @@
+`f`.
+
+# Examples
+```jldoctest
+julia> V = quadratic_space(QQ, 2);
+
+julia> Vf = quadratic_space_with_isometry(V; neg = true);
+
+julia> isometry(Vf)
+[-1    0]
+[ 0   -1]
+```
+"""
+isometry(Vf::QuadSpaceWithIsom) = Vf.f
+
+@doc raw"""
+    order_of_isometry(Vf::QuadSpaceWithIsom) -> IntExt
+
+Given a quadratic space with isometry $(V, f)$, return the order of the
+underlying isometry `f`.
+
+# Examples
+```jldoctest
+julia> V = quadratic_space(QQ, 2);
+
+julia> Vf = quadratic_space_with_isometry(V; neg = true);
+
+julia> order_of_isometry(Vf) == 2
+true
+```
+"""
+order_of_isometry(Vf::QuadSpaceWithIsom) = Vf.n
+
+###############################################################################
+#
+#  Attributes
+#
+###############################################################################
+
+@doc raw"""
+    rank(Vf::QuadSpaceWithIsom) -> Integer
+
+Given a quadratic space with isometry $(V, f)$, return the rank of the underlying
+space `V`.
+
+# Examples
+```jldoctest
+julia> V = quadratic_space(QQ, 2);
+
+julia> Vf = quadratic_space_with_isometry(V; neg = true);
+
+julia> rank(Vf) == 2
+true
+```
+"""
+rank(Vf::QuadSpaceWithIsom) = rank(space(Vf))::Integer
+
+@doc raw"""
+    dim(Vf::QuadSpaceWithIsom) -> Integer
+
+Given a quadratic space with isometry $(V, f)$, return the dimension of the
+underlying space of `V`.
+
+# Examples
+```jldoctest
+julia> V = quadratic_space(QQ, 2);
+
+julia> Vf = quadratic_space_with_isometry(V; neg = true);
+
+julia> dim(Vf) == 2
+true
+```
+"""
+dim(Vf::QuadSpaceWithIsom) = dim(space(Vf))::Integer
+
+@doc raw"""
+    characteristic_polynomial(Vf::QuadSpaceWithIsom) -> QQPolyRingElem
+
+Given a quadratic space with isometry $(V, f)$, return the characteristic
+polynomial of the underlying isometry `f`.
+
+# Examples
+```jldoctest
+julia> V = quadratic_space(QQ, 2);
+
+julia> Vf = quadratic_space_with_isometry(V; neg = true);
+
+julia> characteristic_polynomial(Vf)
+x^2 + 2*x + 1
+```
+"""
+characteristic_polynomial(Vf::QuadSpaceWithIsom) = characteristic_polynomial(isometry(Vf))::QQPolyRingElem
+
+@doc raw"""
+    minimal_polynomial(Vf::QuadSpaceWithIsom) -> QQPolyRingElem
+
+Given a quadratic space with isometry $(V, f)$, return the minimal
+polynomial of the underlying isometry `f`.
+
+# Examples
+```jldoctest
+julia> V = quadratic_space(QQ, 2);
+
+julia> Vf = quadratic_space_with_isometry(V; neg = true);
+
+julia> minimal_polynomial(Vf)
+x + 1
+```
+"""
+minimal_polynomial(Vf) = minimal_polynomial(isometry(Vf))::QQPolyRingElem
+
+@doc raw"""
+    gram_matrix(Vf::QuadSpaceWithIsom) -> QQMatrix
+
+Given a quadratic space with isometry $(V, f)$, return the Gram matrix
+of the underlying space `V` with respect to its standard basis.
+
+# Examples
+```jldoctest
+julia> V = quadratic_space(QQ, 2);
+
+julia> Vf = quadratic_space_with_isometry(V; neg = true);
+
+julia> is_one(gram_matrix(Vf))
+true
+```
+"""
+gram_matrix(Vf::QuadSpaceWithIsom) = gram_matrix(space(Vf))::QQMatrix
```

Comment:
Why all those annotations?

Reply:
Isn't it better for type stability ?

Reply:
Hm, it should not matter, since `gram_matrix` of a quadratic space should be type stable.